### PR TITLE
Refactoring HMC API

### PR DESCRIPTION
**Pull request by xukai92**
Addressing https://github.com/yebai/Turing.jl/issues/431

Still in working.

### CODE REVIEW

**Comment by xukai92** [comment]
https://github.com/yebai/Turing.jl/pull/436/commits/2a4dfb3354aed11423bc3733990638b44da1c62a solves https://github.com/yebai/Turing.jl/issues/434#issuecomment-381111700

**Comment by xukai92** [comment]
@yebai I've refactored core HMC code. Basically I added three functions `_leapfrog`, `_find_H` and `_sample_momentum` and make the original corresponding functions calling them by wrapping Turing.jl internal things. Can you have a look at current change? and let's discuss what's the best way to write the unit tests, e.g. do we simply write another non-Turing.jl based HMC and test them? or something else.

**Comment by yebai** [comment]
The code looks good to me. For unit tests, perhaps we can write 2-3 models without constrained parameters (e.g. gdemo; Bayesian logistic regression, stochastic volatility from the NUTS paper), 

- implement them in both Turing and plain Julia (using AD to compute gradients for the plain Julia version of the models)
- run Turing's HMC sampler on both the Turing version and Julia version

This would allow us to check whether there are errors in the compiler, transformation or gradients. After this step is done, we can focus on debugging the HMC sampler itself by running it on the plain Julia version of models.

**Comment by xukai92** [comment]
Just a small update:

I'm working on this PR today. It turns out that rather than only having Turing-free HMC core functions, we should actually have Turing-free complete HMC step first. Then handling with `vi`, `spl` and `model` outside this level of abstraction (at the moment I think it's hard to do that on the whole HMC level). I'm working on this now.

**Comment by xukai92** [comment]
@yebai So now we have a simple Gaussian and a simple Bayesian linear regression. I'm going to implement a stochastic volatility today or tomorrow. But for the existing two without constrained variables, how should we really debug them?

**Comment by xukai92** [comment]
We actually want to implement a NUTS in the same way like `_hmc_step()` do we?

**Comment by yebai** [comment]
> But for the existing two without constrained variables, how should we really debug them?

@xukai92  It's fine to first test HMC and NUTS on models without constrained variables, just to verify dual averaging and leap-frog integrator is correct. We can change the prior to truncated Gaussian when the initial tests are passed.

**Comment by yebai** [comment]
> We actually want to implement a NUTS in the same way like _hmc_step() do we?

Yes, we would like to be able to unit test NUTS using the same set of models as HMC.

**Comment by xukai92** [comment]
- [ ] NUTS Turing.jl independent implementation

**Comment by xukai92** [comment]
@yebai FYI: I finished my new NUTS implementation before getting all the unit tests done. It turned out that the prvious 2 models to debug on adapation are both OK now, desipte the fact that the initialization sill matters.

**Comment by xukai92** [comment]
Also, I removed the `τ_valid > 0` conditon you added before because, with it, the algorithm can be stuck with a bad (large) step size almost forever.

**Comment by xukai92** [comment]
AppVeyor seems not working

**Comment by xukai92** [comment]
Sorry it's fine after merging with master.

**Comment by yebai** [comment]
> @yebai FYI: I finished my new NUTS implementation before getting all the unit tests done. It turned out that the prvious 2 models to debug on adapation are both OK now, desipte the fact that the initialization sill matters.

Good to know! DO you know what went wrong before?

> Also, I removed the τ_valid > 0 conditon you added before because, with it, the algorithm can be stuck with a bad (large) step size almost forever.

That's OK!

**Comment by xukai92** [comment]
From my experience, bad ininitialization sometimes kills the whole sampling. I think it would be nice to fix that in the PR as well. Do you have any idea how should I do test on that?

Besides, other than this this PR is almost finished. Maybe you can start reviewing - I'm happy to clean up code and add more comments as request. Also, during this refactoring, I feel it might be good to use a conversion that add an underscore "_" before internal functions (i.e. functions not exported). This would help us when writing codes as well. How do you think?

File: src/trace/taskcopy.jl [hunk 1]
```diff
@@ -1,4 +1,3 @@
-include(Pkg.dir("Turing")*"/deps/deps.jl")
```
Comment (xukai92) [inline src/trace/taskcopy.jl]: Removing this makes Turing.jl unable to compile. Why do we want it?

Comment (yebai) [reply src/trace/taskcopy.jl]: This is due to a mistake, will remove this commit tomorrow.

File: src/samplers/support/hmc_core.jl [hunk 2]
```diff
@@ -75,18 +108,52 @@ end
...
-  H
 end
 
+function _find_H(theta::T, p::Vector, lp::Real, stds::Vector) where {T<:Union{Vector,SubArray}}
+
+  p_orig = p .* stds
+
+  H = 0.5 * dot(p_orig, p_orig) + (-lp)
+
+  H = isnan(H) ? Inf : H
```
Comment (willtebbutt) [inline src/samplers/support/hmc_core.jl]: Why is this necessary? Looks like this treats the symptoms of an underlying problem, rather than addressing them directly.

File: test/hmc_core.jl/baeys_lr.jl [hunk 3]
```diff
@@ -0,0 +1,109 @@
+using Distributions, DiffBase
```
Comment (willtebbutt) [inline test/hmc_core.jl/baeys_lr.jl]: Please fix typo in file name.

**Comment by willtebbutt** [comment]
Above review undertaken with @wesselb :)

**Comment by xukai92** [comment]
Thanks @willtebbutt and @wesselb - address them soon.

File: src/samplers/hmcda.jl [hunk 4]
```diff
@@ -82,35 +82,72 @@ function step(model, spl::Sampler{HMCDA}, vi::VarInfo, is_first::Bool)
...
 
-    dprintln(2, "sampling momentum...")
-    p = sample_momentum(vi, spl)
+    lj_func(theta) = begin
 
-    dprintln(2, "recording old values...")
-    old_θ = realpart(vi[spl]); old_logp = getlogp(vi)
-    old_H = find_H(p, model, vi, spl)
+        vi[spl][:] = theta[:]
+        realpart(runmodel(model, vi, spl).logp)
```
[the remark this answers is not in the document]
Comment (xukai92) [reply src/samplers/hmcda.jl]: This is not necessary here. I will remove it.

File: src/samplers/support/hmc_core.jl [hunk 5]
```diff
@@ -11,58 +11,91 @@ end
...
   θ = realpart(_θ)
-  if ADBACKEND == :forward_diff
-    vi[spl] = θ
-    grad = gradient(vi, model, spl)
-  elseif ADBACKEND == :reverse_diff
-    grad = gradient_r(θ, vi, model, spl)
+
+  lp_grad_func(θ::T) where {T<:Union{Vector,SubArray}} = begin
+
+    if ADBACKEND == :forward_diff
```
[the remark this answers is not in the document]
Comment (xukai92) [reply src/samplers/support/hmc_core.jl]: Agree. But it might not be possible to make it into a function because the main reason to introduce this PR and hence this function is to avoid having `vi` in any function signature. But surely I can use a macro to achieve this pattern - will do it for the DRY principle.

File: src/samplers/support/hmc_core.jl [hunk 6]
```diff
@@ -11,58 +11,91 @@ end
...
+      grad = gradient_r(θ, vi, model, spl)
+    end
+
+    return getlogp(vi), grad
+
   end
+
+  rev_func(θ_old::T, old_logp::R) where {T<:Union{Vector,SubArray},R<:Real} = begin
+
+    if ADBACKEND == :forward_diff
```
[the remark this answers is not in the document]
Comment (xukai92) [reply src/samplers/support/hmc_core.jl]: Will solve the same problems using macros.

File: src/samplers/hmcda.jl [hunk 7]
```diff
@@ -148,3 +188,42 @@ function step(model, spl::Sampler{HMCDA}, vi::VarInfo, is_first::Bool)
...
+τ = max(1, round(Int, λ / ϵ))
+dprint(2, "leapfrog for $τ steps with step size $ϵ")
+θ_new, p_new, τ_valid = _leapfrog(θ, p, τ, ϵ, grad_func; rev_func=rev_func, log_func=log_func)
+
+dprint(2, "computing new H...")
+lj_new = lj_func(θ_new)
+H_new = (τ_valid == 0) ? Inf : _find_H(θ_new, p_new, lj_new, stds)
+
+dprint(2, "computing accept rate α...")
+α = min(1, exp(-(H_new - H)))
```
[the remark this answers is not in the document]
Comment (xukai92) [reply src/samplers/hmcda.jl]: Good point. The condition should be `log(u) + H_new < min(H, H_new)`, where `u ~ Uniform(0, 1)`. I don't think we need any Exponential distribution here.

Also I'm going to make this MH criteria into a function.

File: src/samplers/support/hmc_core.jl (hunk 2 above)
Comment (xukai92) [reply src/samplers/support/hmc_core.jl]: The program may give a `NaN` if any constraint applied in the variables is violated or any function inside has some invalid computation, at some point `theta` in the sampling space. To correctly sample in this situation we can simply do rejection sampling, i.e. rejecting any `theta` which gives `NaN`. Plugging rejection sampling in HMC still gives a valid algorithm. One way to do this rejection is simply setting `H` to `Inf`, in which case the MH criteria will reject the sample with a probability of 1.

This is also what Stan's implementation on such rejection: https://github.com/stan-dev/stan/blob/a79d22c3bc9dbfe002ea7174a0f9f8d690e33580/src/stan/mcmc/hmc/static/base_static_hmc.hpp#L61

Comment (wesselb) [reply src/samplers/support/hmc_core.jl]: That makes sense. Thanks for the clarification, @xukai92!

**Comment by xukai92** [comment]
Previous comments are resolved.

**Comment by yebai** [comment]
@willtebbutt could you take a look before I merge this PR? Thanks!

**Comment by yebai** [comment]
> Previous comments are resolved.

Thanks, Kai!

**Comment by willtebbutt** [comment]
Thanks for addressing our concerns @xukai92, just one remaining point. Are the `@gen_local_grad_func` (and related) macros really necessary? Would closures not suffice? i.e.
```julia
function gen_grad_func(vi, spl, model)
    return function (θ::AbstractVector)
        if ADBACKEND == :forward_diff
            vi[spl] = θ
            grad = gradient(vi, model, spl)
        elseif ADBACKEND == :reverse_diff
            grad = gradient_r(θ, vi, model, spl)
        else
            error("An appropriate error.")
        end
        return getlogp(vi), grad
    end
end
```
It's really a matter of taste, but my impression is that a good general rule is to only use metaprogramming where it's really, essentially just because it's tricky to read.

**Comment by xukai92** [comment]
I've replaced macros with closures. Thanks for pointing this - I agree we'd better use closures here.

**Comment by willtebbutt** [comment]
Apologies for the delay. Will review the detail of the NUTS implementation tomorrow morning and, assuming that everything looks fine, I'm happy for this to be merged. (Wessel and I had to refresh our memories regarding NUTS, hence the delay)

File: src/samplers/nuts.jl [hunk 8]
```diff
@@ -132,47 +111,102 @@ function step(model::Function, spl::Sampler{NUTS}, vi::VarInfo, is_first::Bool)
...
-                    model::Function, spl::Sampler, vi::VarInfo)
+"""
+    build_tree(θ::T, r::Vector, logu::Float64, v::Int, j::Int, ϵ::Float64, H0::Float64,
+               lj_func::Function, grad_func::Function, stds::Vector) where {T<:Union{Vector,SubArray}}
+
+Recursively build balanced tree.
+
+Ref: Algorithm 6 on http://www.stat.columbia.edu/~gelman/research/published/nuts.pdf
+"""
+function _build_tree(θ::T, r::Vector, logu::Float64, v::Int, j::Int, ϵ::Float64, H0::Float64,
```
[the remark this answers is not in the document]
Comment (xukai92) [reply src/samplers/nuts.jl]: Make sense! Thanks for pointing out this!

File: src/samplers/nuts.jl [hunk 9]
```diff
@@ -132,47 +111,102 @@ function step(model::Function, spl::Sampler{NUTS}, vi::VarInfo, is_first::Bool)
...
+
+  d = length(θ)
+  r0 = randn(d)
+  H0 = _find_H(θ, r0, lj_func, stds)
+  logu = log(rand()) + -H0
+
+  θm = θ; θp = θ; rm = r0; rp = r0; j = 0; θ_new = θ; n = 1; s = 1
+  local da_stat
+
+  while s == 1 && j <= 5
```
[the remark this answers is not in the document]
Comment (xukai92) [reply src/samplers/nuts.jl]: In practise we have to set such a condition for `j` otherwise the sampler can go super slow if the step-size happens to be very small. Stan uses 10 (maximum 2^10 evaluations) as default and 5 (maximum 2^5 evaluations) was originally chosen by Hong because we were bit slow back then. 

Let me make this number an optional argument in the function. We probably need another interface change to support a user specified maximum `j`.

**Comment by yebai** [comment]
@willtebbutt @xukai92 @wesselb many thanks for the hard work!
